esm_runscripts/compute.py

```diff
@@ -505,16 +507,24 @@ def copy_tools_to_thisrun(config):
 
     # check for recursive creation of the file tree. This prevents the risk of
     # creating a run directory tree inside the `scriptsdir`
-    scriptsdir_inside_fromdir = pathlib.Path(fromdir) in pathlib.Path(scriptsdir).parents
-    if scriptsdir_inside_fromdir and config["general"]["verbose"]:
+    # example:
+    # fromdir:    /mnt/lustre02/work/project/username/basedir
+    # scriptsdir: /mnt/lustre02/work/project/username/basedir/expid/scripts
+    # it is assumed that if 5th parent of the scriptsdir is inside the fromdir
+    # then probably some sort of infinite recursion is entered. These lines
+    # protect such problems
+    scriptsdir_deep_parents = list(pathlib.Path(scriptsdir).parents)[5:]
+    deep_nesting_found = pathlib.Path(expdir) in scriptsdir_deep_parents
+    if deep_nesting_found and config["general"]["verbose"]:
```

Comment:
Isn't `deep_nesting_found = pathlib.Path(expdir) in scriptsdir_deep_parents` always gonna be `true`? I mean `expdir` (`<EXP_DIR>`) is always gonna be parent of `scriptsdir` (`<EXP_DIR>/scripts`) right?

Also I still don't understand this new approach, the question we are trying to answer here is "is the script given in the user command the one inside the scripts folder in the experiment?" right?

In that case, I don't understand why we need to check for nesting paths and parents. I know it is related to the infinite loop issue you mentioned before.

But let's say for example you have your `scriptsdir_deep_parents` be (that's the case for the ALEPH computer):
```
PosixPath('/scratch/awi/mandresm/test/scripts') -> 0
PosixPath('/scratch/awi/mandresm/test') -> 1
PosixPath('/scratch/awi/mandresm') -> 2
PosixPath('/scratch/awi') -> 3
PosixPath('/scratch') -> 4
PosixPath('/') -> 5
```

Then the problem would still remain right?

Reply:
Hi @mandresm,

I don't know if I get this correctly but here is an analogy:

You can think of 5 here as a _maximum iteration count_ in numerical models. If any childhood is met after the 5th level it means that something strange is going on since the runscript is copied to a very deep location. But in the normal case this does not happen. Eg. convergence is met before the max iteration number. 

Here is a code snippet that you can try in IPython:

```python
import pathlib

# scenario 1: normal case, no recursion 
expdir = pathlib.Path("/mnt/lustre02/work/ab0995/a271096/runscripts/issue_156")
scrdir = pathlib.Path("/mnt/lustre02/work/ab0995/a271096/runscripts/issue_156/scripts/")

print(expdir in list(scrdir.parents))        # True
print(expdir in list(scrdir.parents)[5:])    # False

# scenario 2: recursive error is met for some crazy reason
scrdir = pathlib.Path("/mnt/lustre02/work/ab0995/a271096/runscripts/issue_156/scripts/runscripts/issue_156/scripts/runscripts/issue_156/scripts/")
print(expdir in list(scrdir.parents))       # True
print(expdir in list(scrdir.parents)[5:])   # True

# It becomes True and guard is activated because

list(scrdir.parents)[5:]
[PosixPath('/mnt/lustre02/work/ab0995/a271096/runscripts/issue_156/scripts'),
 PosixPath('/mnt/lustre02/work/ab0995/a271096/runscripts/issue_156'),
 PosixPath('/mnt/lustre02/work/ab0995/a271096/runscripts'),
 PosixPath('/mnt/lustre02/work/ab0995/a271096'),
 PosixPath('/mnt/lustre02/work/ab0995'),
 PosixPath('/mnt/lustre02/work'),
 PosixPath('/mnt/lustre02'),
 PosixPath('/mnt'),
 PosixPath('/')]
 
print(fromdir)
PosixPath('/mnt/lustre02/work/ab0995/a271096/runscripts')
```

I had to implement this check because the infinite recursion is really nasty. It creates multiple instances of `esm_runscripts` and you need to use `killall` command to kill them all since it will keep spawning new instances 😱 , meanwhile you think that your simulation is running.

Reply:
After the discussion, I realized that this fix can be done a better way. So, I removed the `deep_nesting_found` in the if statement and added a call to `esm_parser.user_error()` to exit elegantly. 

It is still important to note that when deep recursive occurs there will be multiple instances of `esm_runscripts` running in the background. Therefore, maybe a further call to `os.kill()` can be added. But I think it is Ok for now. As `esm_runscripts` become more robust, we won't run into that recursion issue.